src/toil/leader.py

```diff
@@ -488,28 +489,29 @@ def checkForDeadlocks(self):
         """
         Checks if the system is deadlocked running service jobs.
         """
-        # If there are no updated jobs and at least some jobs issued
-        if len(self.toilState.updatedJobs) == 0 and self.getNumberOfJobsIssued() > 0:
-
-            # If all scheduled jobs are services
-            assert self.serviceJobsIssued + self.preemptableServiceJobsIssued <= self.getNumberOfJobsIssued()
-            if self.serviceJobsIssued + self.preemptableServiceJobsIssued == self.getNumberOfJobsIssued():
-
-                # Sanity check that all issued jobs are actually services
-                for jobNode in self.jobBatchSystemIDToIssuedJob.values():
-                    assert jobNode.jobStoreID in self.toilState.serviceJobStoreIDToPredecessorJob
-
-                # An active service job is one that is not in the process of terminating
-                activeServiceJobs = filter(lambda x : self.serviceManager.isActive(x), self.jobBatchSystemIDToIssuedJob.values())
-
-                # If all the service jobs are active then we have a potential deadlock
-                if len(activeServiceJobs) == len(self.jobBatchSystemIDToIssuedJob):
-                    # We wait self.config.deadlockWait seconds before declaring the system deadlocked
-                    if self.potentialDeadlockedJobs != activeServiceJobs:
-                        self.potentialDeadlockedJobs = activeServiceJobs
-                        self.potentialDeadlockTime = time.time()
-                    elif time.time() - self.potentialDeadlockTime >= self.config.deadlockWait:
-                        raise DeadlockException("The system is service deadlocked - all issued jobs %s are active services" % self.getNumberOfJobsIssued())
+        totalRunningJobs = len(self.batchSystem.getRunningBatchJobIDs())
+        totalServicesIssued = self.serviceJobsIssued + self.preemptableServiceJobsIssued
+        # If there are no updated jobs and at least some jobs running
+        if totalServicesIssued >= totalRunningJobs and len(self.toilState.updatedJobs) == 0 and totalRunningJobs > 0:
+            serviceJobs = filter(lambda x : isinstance(x, ServiceJobNode), self.jobBatchSystemIDToIssuedJob.values())
+            runningServiceJobs = set(filter(lambda x : self.serviceManager.isRunning(x), serviceJobs))
+
+            # If all the running jobs are active services then we have a potential deadlock
```

Comment:
Can we add a sanity check here for ``len(runningServiceJobs) <= totalRunningJobs``?